Nonprofit-Exchange-Hub/web-app
# Add encryption for user password. Add method to search for user by email.

Alright, finally a small PR with a simple change! 

Problem: Real user passwords being stored in database is bad practice.
Solution: Hash the passwords on user creation and only store the hashed version. Validate a password on user login.

We are using [bcrypt](https://www.npmjs.com/package/bcrypt) to hash and validate the passwords. Upon user creation, the password is hashed and replaced with the hashed version for insertion into database.  When a user logs in, their submitted password is hashed and compared against the stored hashed password in the database, which is retrieved by searching for a user with a matching email. If they match, the user entity is returned without a password property. If the password is invalid `{
  "status": 401,
  "error": "Invalid password"
}` is returned. If the email is not in our database `{
  "status": 404,
  "error": "Email not found"
}` is returned. 

@jd2rogers2 I know you were talking about having a more generic response to not reveal if a user exists with a given email, but I have left it more specific for now while we are building it to help with trouble shooting. Shouldn't be too hard to make it more generic later. 

Emails are now unique in our database.

Assuming an empty database and not having bcrypt installed.
1.  `npm install` to get `bcrypt` and `@types/bcrypt` installed.
2. You may need to delete `/server/dist` directory and restart server. I did.
3. You'll probably need to drop the existing database schema because the email column of the users table is now unique. The correct version should be automatically created on server restart.
4. Create new ENV variable `BCRYPT_WORK_FACTOR=10`. This is essentially how difficult the hashed password would be to reverse engineer. Higher number means more secure and takes longer to perform. 10 was given in the NestJS Docs, I've seen 12 used pretty often as well. 
5. Create a new user:
```
curl --request POST \
  --url http://localhost:3001/api/users \
  --header 'Content-Type: application/json' \
  --data '{
	"first_name": "TestFirst",
	"last_name": "TestLast",
	"email": "test@test.com",
	"password": "testpassword"
	}'
```
6. Confirm that the password saved is not the real password, but a hashed version that looks like: `$2b$10$s1TpKNUG3LKpysOTw3VrLO9vNnH0vtK6tPsychpfku3YKlIIFus5.`
7. Confirm that logging in with valid credentials gets you an access token:
```
curl --request POST \
  --url http://localhost:3001/api/auth/login \
  --header 'Content-Type: application/json' \
  --data '{
	"email": "test@test.com",
	"password": "testpassword"
}'
```
Should return something like: `{
  "access_token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpYXQiOjE2MjYxMzczMjIsImV4cCI6MTYyNjEzNzM4Mn0.EVpkoRXRCO1MIOYt9F5dj5yYrRkYXyDq3BNtj5iIgQU"
}`
8. Confirm that wrong email return error:
```
curl --request POST \
  --url http://localhost:3001/api/auth/login \
  --header 'Content-Type: application/json' \
  --data '{
	"email": "wrong@test.com",
	"password": "testpassword"
}'
```
Returns `{
  "status": 404,
  "error": "Email not found"
}`
9. Confirm that wrong password returns error:
```
curl --request POST \
  --url http://localhost:3001/api/auth/login \
  --header 'Content-Type: application/json' \
  --data '{
	"email": "test@test.com",
	"password": "wrong"
}'
```
Returns `{
  "status": 401,
  "error": "Invalid password"
}`

This PR does not address error handling in the user routes outside of findByEmail. Next PR will be for making routes provide better information beyond a 200 status code if there were no server issues.
